# Determining Most Popular Car Brands On Ebay
This dataset was taken from [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data).

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [2]:
print(autos.info())
print("=============================")
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

## Clean Columns
Though most of the data is complete, some columns contain a significant number of null values (in the thousands at least). Also, the column names use camelcase instead of Python's prefered snakecase.

In [3]:
autos.rename(columns={"dateCrawled":"date_crawled",
                      "offerType":"offer_type",
                      "vehicleType":"vehicle_type",
                      "yearOfRegistration":"registration_year",
                      "powerPS":"power_ps",
                      "monthOfRegistration":"registration_month",
                      "fuelType":"fuel_type",
                      "notRepairedDamage":"unrepaired_damage",
                      "dateCreated":"ad_created",
                      "nrOfPictures":"nr_of_pictures",
                      "postalCode":"postal_code",
                      "lastSeen":"last_seen"}, inplace=True)

The column names using camelcase were changed to snakecase to make it more consistent with traditional python naming conventions. 

In [4]:
print(autos.describe())
autos.head()

       registration_year      power_ps  registration_month  nr_of_pictures  \
count       50000.000000  50000.000000        50000.000000         50000.0   
mean         2005.073280    116.355920            5.723360             0.0   
std           105.712813    209.216627            3.711984             0.0   
min          1000.000000      0.000000            0.000000             0.0   
25%          1999.000000     70.000000            3.000000             0.0   
50%          2003.000000    105.000000            6.000000             0.0   
75%          2008.000000    150.000000            9.000000             0.0   
max          9999.000000  17700.000000           12.000000             0.0   

        postal_code  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000  


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

* "nr_of_pictures" column has the value 0 in every row.
* "seller" and "offer_type" have the same value in most rows.

These three columns can be dropped.

In [5]:
autos.drop(columns=["seller", "offer_type", "nr_of_pictures"], inplace=True)
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0  $5,000  control          bus               2004    manuell       158   
1  $8,500  control    limousine               1997  automatik       286   
2  $8,990     test    limousine               2009    manuell       102   
3  $4,350  control   kleinwagen               2007  automatik        71   
4  $1,350     test        kombi               2003    manuell         0   

    model   odometer  registration_month fuel_type       brand  \
0  andere  150,000km                   3       lpg     peugeot   
1     7er  150,000km                   6    benzin         bmw   
2    golf   70,000km                   7    benzin  volkswagen   
3  fortwo   70,000km                   6    benzin       smart   
4   focus  150,000km                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

## Exploring Price And Odometer
"price" and "odometer" are numeric values stored as text.

Let's convert these columns to the numeric dtype.

In [6]:
autos["price"] = autos["price"].str.replace("$", "").str.replace(",", "")
autos["odometer"] = autos["odometer"].str.replace("km", "").str.replace(",", "")

autos["price"] = pd.to_numeric(autos["price"])
autos["odometer"] = pd.to_numeric(autos["odometer"])

autos.rename(columns={"odometer":"odometer_km"}, inplace=True)

In [7]:
autos["price"].value_counts().sort_index(ascending=False)

99999999       1
27322222       1
12345678       3
11111111       2
10000000       1
3890000        1
1300000        1
1234566        1
999999         2
999990         1
350000         1
345000         1
299000         1
295000         1
265000         1
259000         1
250000         1
220000         1
198000         1
197000         1
194000         1
190000         1
180000         1
175000         1
169999         1
169000         1
163991         1
163500         1
155000         1
151990         1
            ... 
66             1
65             5
60             9
59             1
55             2
50            49
49             4
47             1
45             4
40             6
35             1
30             7
29             1
25             5
20             4
18             1
17             3
15             2
14             1
13             2
12             3
11             2
10             7
9              1
8              1
5              2
3              1
2             

Almost half of the total listings are priced at $0. Given that Ebay is an online bidding site, this seems plausible. When examining the higher cost listings, there is a substantial jump from 350,000 to 999,990, and the price jumps from then and onward are inconsistent. Let's remove listings priced above 350,000 totaling 14 / 2357 removed.

In [8]:
autos = autos[autos["price"].between(0, 360000)]
autos["price"].describe()

count     49986.000000
mean       5721.525167
std        8983.617820
min           0.000000
25%        1100.000000
50%        2950.000000
75%        7200.000000
max      350000.000000
Name: price, dtype: float64

## Examining Dates

In [9]:
(autos["date_crawled"]
    .str[:10]
    .value_counts(normalize=True, dropna=False)
    .sort_values()
)

2016-04-07    0.001420
2016-04-06    0.003181
2016-03-18    0.013064
2016-04-05    0.013104
2016-03-06    0.013944
2016-03-13    0.015564
2016-03-05    0.025387
2016-03-24    0.029108
2016-03-16    0.029508
2016-03-27    0.031049
2016-03-17    0.031509
2016-03-25    0.031749
2016-03-31    0.031909
2016-03-10    0.032129
2016-03-23    0.032389
2016-03-26    0.032489
2016-03-11    0.032489
2016-03-22    0.032909
2016-03-09    0.033209
2016-03-08    0.033269
2016-03-30    0.033629
2016-04-01    0.033809
2016-03-15    0.033990
2016-03-29    0.034150
2016-03-28    0.034850
2016-03-19    0.034910
2016-04-02    0.035410
2016-03-07    0.035970
2016-04-04    0.036490
2016-03-14    0.036630
2016-03-12    0.036770
2016-03-21    0.037490
2016-03-20    0.037831
2016-04-03    0.038691
Name: date_crawled, dtype: float64

It seems the crawler started slowing down near the beginning of April.

In [10]:
(autos["ad_created"]
    .str[:10]
    .value_counts(normalize=True, dropna=False)
    .sort_values()
)

2016-01-13    0.000020
2016-02-08    0.000020
2015-09-09    0.000020
2016-01-07    0.000020
2016-01-22    0.000020
2015-06-11    0.000020
2015-12-05    0.000020
2015-08-10    0.000020
2016-02-17    0.000020
2016-01-14    0.000020
2016-02-22    0.000020
2016-02-07    0.000020
2016-02-16    0.000020
2016-02-11    0.000020
2016-01-16    0.000020
2016-01-29    0.000020
2015-11-10    0.000020
2016-02-01    0.000020
2016-01-03    0.000020
2015-12-30    0.000020
2016-02-18    0.000040
2016-02-02    0.000040
2016-02-09    0.000040
2016-02-05    0.000040
2016-02-14    0.000040
2016-01-10    0.000040
2016-02-20    0.000040
2016-02-24    0.000040
2016-02-26    0.000040
2016-02-21    0.000060
                ...   
2016-03-06    0.015124
2016-03-13    0.016925
2016-03-05    0.023046
2016-03-24    0.029088
2016-03-16    0.030008
2016-03-27    0.030909
2016-03-17    0.031189
2016-03-10    0.031869
2016-03-25    0.031889
2016-03-31    0.031909
2016-03-23    0.032189
2016-03-26    0.032569
2016-03-22 

The most number of ads were created around the month of March.

In [11]:
(autos["last_seen"]
    .str[:10]
    .value_counts(normalize=True, dropna=False)
    .sort_values()
)

2016-03-05    0.001080
2016-03-06    0.004421
2016-03-07    0.005362
2016-03-18    0.007422
2016-03-08    0.007582
2016-03-13    0.008983
2016-03-09    0.009843
2016-03-10    0.010763
2016-03-11    0.012524
2016-03-14    0.012804
2016-03-19    0.015744
2016-03-15    0.015884
2016-03-27    0.016024
2016-03-16    0.016445
2016-03-26    0.016965
2016-03-23    0.018585
2016-03-25    0.019205
2016-03-24    0.019565
2016-03-20    0.020706
2016-03-21    0.020726
2016-03-28    0.020846
2016-03-22    0.021586
2016-03-29    0.022326
2016-04-01    0.023106
2016-03-12    0.023807
2016-03-31    0.023827
2016-04-04    0.024627
2016-03-30    0.024847
2016-04-02    0.024887
2016-04-03    0.025367
2016-03-17    0.027928
2016-04-05    0.124275
2016-04-07    0.130957
2016-04-06    0.220982
Name: last_seen, dtype: float64

Around the same time there was a spike in the "last_seen" values, there was also a dip in the "date_crawled" values. It is unlikely that this event correlates with car sales.

## Exploring Registration Year

In [12]:
autos["registration_year"].describe()

count    49986.000000
mean      2005.075721
std        105.727161
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The min and max registration years for the cars are 1000 and 9999 respectively. Given that cars didn't exist in the year 1000 and that 9999 is about 8000 years into the future, it's safe to assume that these values are incorrect.

Lets remove all data that has the registration year before 1900 and after 2016.

In [13]:
autos = autos[autos["registration_year"].between(1900, 2016)]
(autos["registration_year"]
     .value_counts(normalize=True)
     .sort_index()
)

1910    0.000187
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000042
1937    0.000083
1938    0.000021
1939    0.000021
1941    0.000042
1943    0.000021
1948    0.000021
1950    0.000062
1951    0.000042
1952    0.000021
1953    0.000021
1954    0.000042
1955    0.000042
1956    0.000104
1957    0.000042
1958    0.000083
1959    0.000146
1960    0.000687
1961    0.000125
1962    0.000083
1963    0.000187
1964    0.000250
1965    0.000354
1966    0.000458
1967    0.000562
1968    0.000541
          ...   
1987    0.001562
1988    0.002957
1989    0.003770
1990    0.008226
1991    0.007414
1992    0.008122
1993    0.009268
1994    0.013745
1995    0.027324
1996    0.030073
1997    0.042236
1998    0.051087
1999    0.062438
2000    0.069852
2001    0.056273
2002    0.052753
2003    0.056794
2004    0.057002
2005    0.062792
2006    0.056377
2007    0.047984
2008    0.046464
2009    0.043673
2010    0.033260
2011    0.034030
2012    0.027553
2013    0.016786
2014    0.0138

## Analyzing The Top Car Brands
Let's analyze the top 20 most listed car brands.

In [14]:
top_20_brands = (autos["brand"]
    .value_counts()
    .sort_values(ascending=False)
    .head(20)
)
top_20_brands

volkswagen        10185
bmw                5283
opel               5194
mercedes_benz      4579
audi               4149
ford               3350
renault            2274
peugeot            1418
fiat               1242
seat                873
skoda               770
mazda               727
nissan              725
smart               668
citroen             668
toyota              599
sonstige_autos      523
hyundai             473
volvo               444
mini                415
Name: brand, dtype: int64

It looks like "volkswagon" is ahead in popularity of listings by almost twice the amount of the second most popular brand,"bmw". The popularity could be a result of its average price. If it's cheaper than its competition then the expectation of sell numbers would be higher.

Let's analyze the average price of the top 5 of the top 20 car brands.

In [15]:
top_20_avg = {}
top_20_mil = {}

count = 0
for brand in top_20_brands.index:
    row = autos[autos["brand"] == brand]
    avg = row["price"].mean()
    top_20_avg[brand] = int(avg)
    
    mil = row["odometer_km"].mean()
    top_20_mil[brand] = int(mil)
    
    count += 1
    if count == 5:
        break
    
for brand in top_20_avg:
    print(brand + ": " + str(top_20_avg[brand]))

audi: 9093
mercedes_benz: 8485
opel: 2876
bmw: 8102
volkswagen: 5231


As to be expected, volkswagen is cheaper in price than most of the other top cars. However, it's not cheaper than opel which only came in second. Maybe the volkswagen has certain features tied to it that makes it more attractive to buyers. It is not unlikely to think that the volkswagen is riding in a certain "sweet-spot" between cost and brand recognition.

In [16]:
avg_series = pd.Series(top_20_avg)
mil_series = pd.Series(top_20_mil)

price_and_mile = pd.DataFrame(avg_series, columns=["average_price"])
price_and_mile["average_mileage"] = mil_series

price_and_mile

average_price  average_mileage
audi                    9093           129287
bmw                     8102           132431
mercedes_benz           8485           130856
opel                    2876           129223
volkswagen              5231           128724

It remains that the average price is the only differentiating factor between the brands. The average mileage between the brands is too similar to have an affect on their listing popularity.